### ISSUE TITLE
updateParameters() behaves strangely when manually sharing via set() on parameter tensors

### ISSUE THREAD

**Issue by kbullaughey** [torch/nn#395]
I ran into a problem that I'm not sure whether it is a bug or my misuse of `nn`. 

I find it conceptually simpler to manually set up parameter sharing by calling `set()` on parameter tensors I want tied:

    p2:set(p1)
    p3:set(p1)

Where I want p2 and p3 tied to p1

This works fine if I get linearized par and gradPar vectors using `getParameters()` and then manually update my parameter vector. I first get my vectorized parameters:

    par, gradPar = net:getParameters()

And during updates (after forward-backward):

    par:add(gradPar:mul(-learnRate))

However if I use the built-in function:

   net:updateParameters(learnRate)

Then the par vector after I call `updateParameters()` is different than one would expect and convergence is slow and fit is poor.

Here's a full illustration of the problem. First the code:

    require 'nn'
    
    torch.manualSeed(1)
    initialPar = torch.Tensor(12):uniform(-0.1, 0.1)
    x = torch.Tensor(3):uniform(-1, 1)
    y = torch.Tensor(3):uniform(-1, 1)
    outputLine = 1
    
    function makeNet(initialPar)
      local net = {}
      -- I create a simple network that consists of two 3x3 linear maps.
      local g = nn.Sequential()
      g:add(nn.Linear(3,3))
      g:add(nn.Linear(3,3))
      -- Tie parameters of two linear maps together
      local parTensors, gradParTensors = g:parameters()
      parTensors[3]:set(parTensors[1])
      parTensors[4]:set(parTensors[2])
      gradParTensors[3]:set(gradParTensors[1])
      gradParTensors[4]:set(gradParTensors[2])
    
      net.graph = g
      net.par, net.gradPar = g:getParameters()
      net.par:copy(initialPar)
      net.graph:zeroGradParameters()
      net.criterion = nn.MSECriterion()
      return net
    end
    
    function roundTrip(net, x,y)
      net.graph:forward(x)
      net.criterion:forward(net.graph.output, y)
      net.criterion:backward(net.graph.output, y)
      net.graph:backward(x, net.criterion.gradInput)
    end
    
    function log(s)
      print(outputLine .. ": " .. s)
      outputLine = outputLine + 1
    end
    
    function summarize(net, label)
      log("[" .. label .. "] par sum: " .. net.par:sum() .. "; gradPar sum: " .. net.gradPar:sum())
    end
    
    a = makeNet(initialPar)
    b = makeNet(initialPar)
    
    summarize(a, 'a')
    summarize(b, 'b')
    
    roundTrip(a, x, y)
    roundTrip(b, x, y)
    
    summarize(a, 'a')
    summarize(b, 'b')
    
    a.graph:updateParameters(0.05)
    b.par:add(b.gradPar * (-0.05))
    
    summarize(a, 'a')
    summarize(b, 'b')
    
And here's the output:

    1: [a] par sum: -0.1263070480898; gradPar sum: 0	
    2: [b] par sum: -0.1263070480898; gradPar sum: 0	
    3: [a] par sum: -0.1263070480898; gradPar sum: -0.64602592080819	
    4: [b] par sum: -0.1263070480898; gradPar sum: -0.64602592080819	
    5: [a] par sum: -0.061704456008983; gradPar sum: -0.64602592080819	
    6: [b] par sum: -0.094005752049393; gradPar sum: -0.64602592080819	

I would have expected the first number in each of line 5 and 6 to be the same.

**Comment by fmassa** [comment]
The problem arises because when you call `:updateParameters()` it doesn't take into account the sharing of the modules.
What it actually does is that it iterates over all the modules and call `:updateParameters()` for every module.
In your case, it's called twice for the linear, except that the gradients for the linear were already accumulated for both `Linear` (because you shared them).
You can see that if you replace in your example the line where you update the parameters by
```lua
a.graph:updateParameters(0.05)
b.par:add(b.gradPar * (-0.05))
b.par:add(b.gradPar * (-0.05)) -- update parameters is summing twice !
```
The solution is not to share the `gradients` in you model `a`.

I think this issue deserves an explicit mention in the docs, saying that if you are going to use `getParameters`, then you need to share the `gradWeight` and `gradBias` as well as the `weight` and `bias`, and if manually iterating over the modules (like in Facebook's `nn.Optim`), then you shouldn't share the gradients.

**Comment by kbullaughey** [comment]
Hm...you seem to be correct. Although I think I'll stick with sharing my gradient parameters and doing updates manually. It makes more sense to me to have a one-to-one correspondence between my linearized parameter vector and my linearized gradient wrt parameters vector. For example, this makes it easier to track where vanishing or exploding gradients might be happening, etc.

But given this doesn't seem to be a bug, I'll close this issue.

**Comment by kbullaughey** [comment]
Thanks very much for your help! It prompted me to dig into the code some more too, which was also helpful.
